### PR TITLE
chore(linter): add golangci-lint tool for checking the gocode base

### PR DESCRIPTION
**Pull request by mittachaitu**
**Why is this PR required? What issue does it fix?**:
This PR is required to make static analysis of the codebase and
it is achieved by adding lint tool in the codebase.

**What this PR does?**:
This PR adds [golangci-lint](https://golangci-lint.run/usage/configuration/) tool to verify linting issues in the codebase

**Does this PR require any upgrade changes?**:
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:
Yes

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

- Currently, disabled `structcheck` analysis due to some [false-positive](https://github.com/golangci/golangci-lint/issues/537) errors and it can be enabled once the issue was resolved.
- Disabled linter on test directory as it contains repeated code(Need to fix it).

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

### CODE REVIEW

**Comment by mittachaitu** [comment]
Run the command `make golint` on my system without any changes to code base
```sh
sai@sai:~/gocode/src/github.com/openebs/lvm-localpv$ golangci-lint run  -v
INFO [config_reader] Config search paths: [./ /home/sai/gocode/src/github.com/openebs/lvm-localpv /home/sai/gocode/src/github.com/openebs /home/sai/gocode/src/github.com /home/sai/gocode/src /home/sai/gocode /home/sai /home /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck] 
INFO [loader] Go packages loading at mode 575 (imports|types_sizes|compiled_files|deps|name|exports_file|files) took 685.894506ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 9.943742ms 
INFO [linters context/goanalysis] analyzers took 7.652354677s with top 10 stages: buildir: 5.282137223s, S1038: 129.576922ms, directives: 95.33991ms, unused: 88.872358ms, S1039: 75.678618ms, S1007: 60.736466ms, SA9006: 60.189626ms, inspect: 56.910141ms, S1004: 54.17343ms, SA4020: 49.532466ms 
INFO [runner/skip dirs] Skipped 8 issues from dir tests by pattern tests 
INFO [runner/skip dirs] Skipped 1 issues from dir tests/k8svolume by pattern tests 
INFO [runner/max_same_issues] 2/5 issues with text "Error return value of `env.Set` is not checked" were hidden, use --max-same-issues 
INFO [runner] Issues before processing: 55, after processing: 36 
INFO [runner] Processors filtering stat (out/in): max_same_issues: 36/38, max_from_linter: 36/36, path_shortener: 36/36, severity-rules: 36/36, sort_results: 36/36, cgo: 55/55, diff: 38/38, max_per_file_from_linter: 38/38, nolint: 45/45, filename_unadjuster: 55/55, autogenerated_exclude: 46/46, identifier_marker: 46/46, exclude: 46/46, source_code: 36/36, exclude-rules: 45/46, uniq_by_line: 38/45, path_prefixer: 36/36, path_prettifier: 55/55, skip_files: 55/55, skip_dirs: 46/55 
INFO [runner] processing took 6.980144ms with stages: nolint: 3.903771ms, exclude-rules: 1.587935ms, identifier_marker: 535.372µs, path_prettifier: 297.048µs, autogenerated_exclude: 271.3µs, source_code: 236.932µs, skip_dirs: 76.481µs, max_same_issues: 26.215µs, uniq_by_line: 17.989µs, max_from_linter: 5.93µs, max_per_file_from_linter: 5.822µs, path_shortener: 5.713µs, cgo: 5.271µs, filename_unadjuster: 2.697µs, diff: 501ns, exclude: 275ns, skip_files: 270ns, sort_results: 248ns, severity-rules: 231ns, path_prefixer: 143ns 
INFO [runner] linters took 2.86926146s with stages: goanalysis_metalinter: 2.862201902s 
pkg/driver/controller.go:799:6: `validateCapabilities` is unused (deadcode)
func validateCapabilities(caps []*csi.VolumeCapability) bool {
     ^
pkg/driver/driver.go:28:5: `supportedAccessMode` is unused (deadcode)
var supportedAccessMode = &csi.VolumeCapability_AccessMode{
    ^
pkg/usage/usage.go:198:14: Error return value of `v.getVersion` is not checked (errcheck)
	v.getVersion(false)
	            ^
pkg/usage/usage.go:212:14: Error return value of `v.getVersion` is not checked (errcheck)
	v.getVersion(false)
	            ^
pkg/usage/usage.go:258:14: Error return value of `v.getVersion` is not checked (errcheck)
	v.getVersion(override)
	            ^
pkg/usage/versionset.go:60:9: Error return value of `env.Set` is not checked (errcheck)
	env.Set(clusterUUID, v.id)
	       ^
pkg/usage/versionset.go:69:9: Error return value of `env.Set` is not checked (errcheck)
	env.Set(clusterArch, v.k8sArch)
	       ^
pkg/usage/versionset.go:70:9: Error return value of `env.Set` is not checked (errcheck)
	env.Set(clusterVersion, v.k8sVersion)
	       ^
pkg/driver/grpc.go:195:14: Error return value of `server.Serve` is not checked (errcheck)
	server.Serve(listener)
	            ^
pkg/usage/usage.go:100:2: `documentTitle` is unused (structcheck)
	documentTitle string
	^
pkg/usage/usage.go:85:2: `clientID` is unused (structcheck)
	clientID string
	^
pkg/usage/usage.go:88:2: `campaignSource` is unused (structcheck)
	campaignSource string
	^
pkg/usage/usage.go:91:2: `campaignName` is unused (structcheck)
	campaignName string
	^
pkg/usage/usage.go:95:2: `dataSource` is unused (structcheck)
	dataSource string
	^
pkg/usage/usage.go:44:2: `category` is unused (structcheck)
	category string
	^
pkg/usage/usage.go:46:2: `action` is unused (structcheck)
	action string
	^
pkg/usage/usage.go:48:2: `label` is unused (structcheck)
	label string
	^
pkg/usage/usage.go:51:2: `value` is unused (structcheck)
	value int64
	^
pkg/usage/usage.go:67:2: `appVersion` is unused (structcheck)
	appVersion string
	^
pkg/usage/usage.go:70:2: `appInstallerID` is unused (structcheck)
	appInstallerID string
	^
pkg/usage/usage.go:73:2: `appID` is unused (structcheck)
	appID string
	^
pkg/usage/usage.go:76:2: `appName` is unused (structcheck)
	appName string
	^
pkg/driver/controller.go:152:12: S1039: unnecessary use of fmt.Sprintf (gosimple)
		errMsg = fmt.Sprintf("failed lvmvol must have error set")
		         ^
pkg/driver/grpc.go:130:2: S1023: redundant `return` statement (gosimple)
	return
	^
pkg/driver/agent.go:387:5: S1002: should omit comparison to bool constant, can be simplified to `!mount.IsMountPath(path)` (gosimple)
	if mount.IsMountPath(path) == false {
	   ^
pkg/driver/grpc.go:71:5: S1002: should omit comparison to bool constant, can be simplified to `log` (gosimple)
	if log == true {
	   ^
pkg/driver/grpc.go:77:5: S1002: should omit comparison to bool constant, can be simplified to `log` (gosimple)
	if log == true {
	   ^
pkg/lvm/lvm_util.go:221:5: S1002: should omit comparison to bool constant, can be simplified to `resizefs` (gosimple)
	if resizefs == true {
	   ^
pkg/lvm/iolimiter.go:42:2: S1008: should use 'return set' instead of 'if set { return true }; return false' (gosimple)
	if set {
	^
pkg/lvm/lvm_util.go:373:4: ineffectual assignment to err (ineffassign)
			err = fmt.Errorf("invalid format of %v=%v for vg %v: %v", key, m[key], vg.Name, err)
			^
pkg/lvm/lvm_util.go:388:4: ineffectual assignment to err (ineffassign)
			err = fmt.Errorf("invalid format of %v=%v for vg %v: %v", key, m[key], vg.Name, err)
			^
pkg/driver/agent.go:40:2: SA1019: package sigs.k8s.io/controller-runtime/pkg/runtime/signals is deprecated: use pkg/manager/signals instead. (staticcheck)
	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
	^
pkg/driver/controller.go:30:2: SA1019: package sigs.k8s.io/controller-runtime/pkg/runtime/signals is deprecated: use pkg/manager/signals instead. (staticcheck)
	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
	^
pkg/lvm/mount.go:303:2: SA5001: should check returned error before deferring f.Close() (staticcheck)
	defer f.Close()
	^
pkg/mgmt/lvmnode/lvmnode.go:81:6: SA4006: this value of `node` is never used (staticcheck)
		if node, err = nodebuilder.NewKubeclient().WithNamespace(namespace).Create(node); err != nil {
		   ^
pkg/mgmt/lvmnode/lvmnode.go:111:5: SA4006: this value of `node` is never used (staticcheck)
	if node, err = nodebuilder.NewKubeclient().WithNamespace(namespace).Update(node); err != nil {
	   ^
INFO File cache stats: 10 entries of total size 92.4KiB 
INFO Memory: 37 samples, avg is 235.6MB, max is 409.9MB 
INFO Execution took 3.571919128s    
```
--------------------------------------After fixing the lining issues ----------------------
```sh
make golint 
--> Running golint
./buildscripts/linter.sh
Checking linting issues
Completed golint no recommendations !!
--------------------------------
```

File: buildscripts/linter.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+
+# Copyright © 2020 The OpenEBS Authors
```
Comment (kmova) [inline buildscripts/linter.sh:3]: 2021 :-)

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -46,6 +46,10 @@ jobs:
           pattern: '*.sh'
           exclude: './.git/*,./vendor/*'
 
+      - name: Lint Check
+        run: |
+          sudo curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b "${GOPATH}"/bin v1.40.1
```
Comment (pawanpraka1) [inline .github/workflows/build.yml:51]: can we make it part of `make bootstrap`

File: .github/workflows/pull_request.yml [hunk 3]
```diff
@@ -50,7 +50,9 @@ jobs:
         run: make bootstrap
 
       - name: Lint Check
-        run: make golint
+        run: |
+          curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sudo sh -s -- -b "${GOPATH}"/bin v1.40.1
```
Comment (pawanpraka1) [inline .github/workflows/pull_request.yml:54]: same here.

File: buildscripts/linter.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    ## dupl: Tool for code clone detection
+    ## revive: Drop-in replacement of golint. It allows to enable or disable
+    ##         rules using configuration file.
+    ## bodyclose: checks whether HTTP response body is closed successfully
+    ## goconst: Find repeated strings that could be replaced by a constant
+    ## misspell: Finds commonly misspelled English words in comments
+    ##
+    ## NOTE: Disabling structcheck since it is reporting false positive cases
+    ##       for more information look at https://github.com/golangci/golangci-lint/issues/537
+    golangci-lint run -E exportloopref,dupl,revive,bodyclose,goconst,misspell -D structcheck --timeout 5m0s
```
Comment (pawanpraka1) [inline buildscripts/linter.sh:49]: once installation becomes part of bootstrap, we can directly run this in `make golint`. see if we need this script after that?

File: pkg/lvm/lvm_util.go [hunk 5]
```diff
@@ -49,6 +49,8 @@ const (
 	LVExtend = "lvextend"
 
 	PVScan = "pvscan"
+
+	yes = "yes"
```
Comment (pawanpraka1) [inline pkg/lvm/lvm_util.go:53]: use capital letters for constants.

File: .github/workflows/build.yml (hunk 2 above)
Comment (mittachaitu) [reply .github/workflows/build.yml:51]: Do you mean to use `go get github.com/golangci/golangci-lint/cmd/golangci-lint@v1.40.1`? One small gotcha [golangci-lint docs](https://golangci-lint.run/usage/install/#install-from-source) are recommending to install via binary installation mode.

Comment (mittachaitu) [reply .github/workflows/build.yml:51]: So I can think of following way in make file
```sh
## Binaries required for make targets
## NOTE: We can add commands here which can't be installed
## via "go get"
## Example: golangci-lint tool document is not recommending
##			to use `go get <path>`. For more info:
##          https://golangci-lint.run/usage/install/#install-from-source
EXTERNAL_BINARIES=\
	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b "${GOPATH}"/bin v1.40.1

.PHONY: bootstrap
bootstrap: controller-gen
	@for tool in  $(EXTERNAL_BINARIES) ; do \
		echo "+ Installing $$tool" ; \
		$tool \
	done
```
Is it good have something like above?

Comment (pawanpraka1) [reply .github/workflows/build.yml:51]: make another end point for installating golangci and put that as dependent for bootstrap. There we can fire curl command.

**Review by pawanpraka1 — APPROVED** [review]
looks good. Thanks @mittachaitu .

**Pull request merged**